### PR TITLE
feature(U-6932054538): Implement the /api/hasPermission endpoint

### PR DESCRIPTION
**Pull request by romannaglic**
The update includes new and updated dependencies in build.gradle for authentication and permissions-related functionality. Changes to the /api/hasPermission and /login endpoints allow for handling of permissions and user login respectively. New exceptions have been added to handle validation of users, and changes to the UserRepo, User, and Tenant entity classes have been made to accommodate these changes. The schema for the `user` table has also been updated. Unit tests for these new functionalities have also been added.

### CODE REVIEW

**Review by montesmoci — APPROVED** [review]

File: src/test/java/io/unityfoundation/MockAuthenticationProvider.java [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+import reactor.core.publisher.FluxSink;
+
+@Singleton
+public class MockAuthenticationProvider implements AuthenticationProvider<HttpRequest<?>> {
+
+  @Override
+  public Publisher<AuthenticationResponse> authenticate(@Nullable HttpRequest<?> httpRequest,
+      AuthenticationRequest<?, ?> authenticationRequest) {
+    return Flux.create(emitter -> {
+      if ((authenticationRequest.getIdentity().equals("wilsonj@unityfoundation.io")
```
Comment (jrw972) [inline src/test/java/io/unityfoundation/MockAuthenticationProvider.java:20]: Use name@test.test

File: src/main/java/io/unityfoundation/auth/HasPermissionRequest.java [hunk 2]
```diff
@@ -0,0 +1,13 @@
+package io.unityfoundation.auth;
+
+import io.micronaut.serde.annotation.Serdeable;
+import jakarta.validation.constraints.NotNull;
+import java.util.List;
+
+@Serdeable
+public record HasPermissionRequest(
```
Comment (jrw972) [inline src/main/java/io/unityfoundation/auth/HasPermissionRequest.java:8]: I think this needs an extra flag indicating that SUBTENANT scope is allowed.  Or is that what' I'm seeing in the AuthController?

File: src/main/java/io/unityfoundation/auth/AuthController.java [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+    User user = userRepo.findByEmail(authentication.getName()).orElse(null);
+    if (user == null || user.getStatus() != User.UserStatus.ENABLED) {
+      return false;
+    }
+
+    if (requestDTO.tenantId() == null) {
+      return false;
+    }
+
+    Optional<Tenant> t = user.getTenants().stream()
```
Comment (jrw972) [inline src/main/java/io/unityfoundation/auth/AuthController.java:54]: This won't work.  If I'm an auth admin at the SYSTEM level, then the tenant from the request will not be in `user.getTenants`.

**Review by jrw972 — CHANGES_REQUESTED** [review]

File: src/test/java/io/unityfoundation/MockAuthenticationProvider.java (hunk 1 above)
Comment (romannaglic) [reply src/test/java/io/unityfoundation/MockAuthenticationProvider.java]: changed

File: src/main/java/io/unityfoundation/auth/HasPermissionRequest.java (hunk 2 above)
Comment (romannaglic) [reply src/main/java/io/unityfoundation/auth/HasPermissionRequest.java:8]: currently I set subtenant flag as query parameter, but if you think that HasPermissionRequest record would be better position I can change that

File: src/main/java/io/unityfoundation/auth/AuthController.java (hunk 3 above)
Comment (romannaglic) [reply src/main/java/io/unityfoundation/auth/AuthController.java:54]: Logic assumes that there is tenant with name "SYSTEM". That way all logic is much more simple to implement.
The SYSTEM tenant is created with initialization script

File: src/main/java/io/unityfoundation/auth/HasPermissionRequest.java [hunk 4]
```diff
@@ -0,0 +1,13 @@
...
+    @NotNull Long tenantId,
+    @NotNull Long serviceId,
```
Comment (montesmoci) [inline src/main/java/io/unityfoundation/auth/HasPermissionRequest.java:10]: @jrw972 @romannaglic Can serviceId be a String instead? I was hoping to derive the value from application.yml's micronaut.application.name configuration (right now it is 'app' but I intend on changing it to "Libre311").

Comment (jrw972) [reply src/main/java/io/unityfoundation/auth/HasPermissionRequest.java:10]: We could require that service name be unique and take that in the request.

Comment (romannaglic) [reply src/main/java/io/unityfoundation/auth/HasPermissionRequest.java]: serviceId and tenantId chenged to string.

File: src/main/java/io/unityfoundation/auth/UnityAuthenticationProvider.java [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+          } else {
+            return Mono.just(AuthenticationResponse.success((String) authenticationRequest.getIdentity()));
+          }
+        });  }
+
+  private AuthenticationFailed validate(User user,
+      AuthenticationRequest<?, ?> authenticationRequest) {
+    AuthenticationFailed authenticationFailed = null;
+    if (user == null) {
+      authenticationFailed = new AuthenticationFailed(USER_NOT_FOUND);
```
Comment (jrw972) [inline src/main/java/io/unityfoundation/auth/UnityAuthenticationProvider.java:58]: For security, use `CREDENTIALS_DO_NOT_MATCH` so we don't leak who has an account and who does not.

File: src/main/resources/db/migration/V1__initial_schema.sql [hunk 6]
```diff
@@ -1,51 +1,61 @@
...
 );
+ALTER TABLE tenant_service
+    ADD CONSTRAINT tenant_service_tenant_FK FOREIGN KEY (tenant_id) REFERENCES tenant (id);
+ALTER TABLE tenant_service
+    ADD CONSTRAINT tenant_service_service_FK FOREIGN KEY (service_id) REFERENCES service (id);
 
 CREATE TABLE permission
 (
-    id          int AUTO_INCREMENT PRIMARY KEY,
+    id          bigint AUTO_INCREMENT PRIMARY KEY,
```
Comment (jrw972) [inline src/main/resources/db/migration/V1__initial_schema.sql:31]: Name should be unique.

File: src/main/resources/db/migration/V1__initial_schema.sql [hunk 7]
```diff
@@ -1,51 +1,61 @@
 CREATE TABLE tenant
 (
-    id     int AUTO_INCREMENT PRIMARY KEY,
-    name   varchar(255) NOT NULL,
-    status varchar(255) NOT NULL
+    id          bigint AUTO_INCREMENT PRIMARY KEY,
+    name        varchar(255) NOT NULL,
```
Comment (jrw972) [inline src/main/resources/db/migration/V1__initial_schema.sql:4]: Name should be unique.

File: src/main/resources/db/migration/V1__initial_schema.sql [hunk 8]
```diff
@@ -1,51 +1,61 @@
...
+    description text,
+    status      varchar(255) NOT NULL
 );
 
 CREATE TABLE service
 (
-    id          int AUTO_INCREMENT PRIMARY KEY,
+    id          bigint AUTO_INCREMENT PRIMARY KEY,
```
Comment (jrw972) [inline src/main/resources/db/migration/V1__initial_schema.sql:11]: `name` should be unique.

File: src/main/resources/db/migration/V1__initial_schema.sql [hunk 9]
```diff
@@ -1,51 +1,61 @@
...
     name        varchar(255) NOT NULL,
     description text,
     scope       varchar(255)
 );
 
 CREATE TABLE role
 (
-    id          int AUTO_INCREMENT PRIMARY KEY,
+    id          bigint AUTO_INCREMENT PRIMARY KEY,
```
Comment (jrw972) [inline src/main/resources/db/migration/V1__initial_schema.sql:39]: Name should be unique.

File: src/main/resources/db/migration/V1__initial_schema.sql [hunk 10]
```diff
@@ -1,51 +1,61 @@
...
+ALTER TABLE role_permission
+    ADD CONSTRAINT role_permission_permission_FK FOREIGN KEY (permission_id) REFERENCES permission (id);
+ALTER TABLE role_permission
+    ADD CONSTRAINT role_permission_role_FK FOREIGN KEY (role_id) REFERENCES role (id);
+
 
 CREATE TABLE user
 (
-    id       int AUTO_INCREMENT PRIMARY KEY,
+    id       bigint AUTO_INCREMENT PRIMARY KEY,
```
Comment (jrw972) [inline src/main/resources/db/migration/V1__initial_schema.sql:58]: email should be unique.

**Review by jrw972 — CHANGES_REQUESTED** [review]

File: src/main/java/io/unityfoundation/auth/UnityAuthenticationProvider.java (hunk 5 above)
Comment (romannaglic) [reply src/main/java/io/unityfoundation/auth/UnityAuthenticationProvider.java]: changed

File: src/main/resources/db/migration/V1__initial_schema.sql (hunk 6 above)
Comment (romannaglic) [reply src/main/resources/db/migration/V1__initial_schema.sql:33]: added unique constraint

File: src/main/resources/db/migration/V1__initial_schema.sql (hunk 8 above)
Comment (romannaglic) [reply src/main/resources/db/migration/V1__initial_schema.sql:12]: added unique constraint

File: src/main/resources/db/migration/V1__initial_schema.sql (hunk 9 above)
Comment (romannaglic) [reply src/main/resources/db/migration/V1__initial_schema.sql:42]: added unique constraint

File: src/main/resources/db/migration/V1__initial_schema.sql (hunk 10 above)
Comment (romannaglic) [reply src/main/resources/db/migration/V1__initial_schema.sql:62]: added unique constraint

**Review by jrw972 — APPROVED** [review]

**Pull request merged**